functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <example_srvs/srv/get_path.hpp>
+#include <example_srvs/srv/set_map.hpp>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <std_msgs/msg/u_int8_multi_array.hpp>
+
+using namespace std::chrono_literals;
+
+/**
+ * @brief Represents a position in the map
+ */
+struct Position {
+  size_t x;
+  size_t y;
+};
+
+// Operator overload for position comparison
+bool operator==(Position const& lhs, Position const& rhs) {
+  return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+using Path = std::vector<Position>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, Path const& path) {
+  for (auto const& pose : path) {
+    os << "(" << pose.x << ", " << pose.y << ")\n";
+  }
+  return os;
+}
+
+template <typename T>
+class Map {
+ public:
+  Map(){};
+  /**
+   * @brief Construct a map from a 2D vector
+   * @param data The 2D vector to construct the map from
+   */
+  Map(std::vector<std::vector<T>> data) : data_{data} {};
+  T at(Position const& pos) const { return data_.at(pos.y).at(pos.x); }
+  std::vector<std::vector<T>>& get_data() { return data_; }
+
+ private:
+  std::vector<std::vector<T>> data_;
+};
+
+struct PathGenerator {
+  struct MiddlewareHandle {
+    // Define map service callback type
+    using SetMapServiceCallback = std::function<void(
+        std::shared_ptr<example_srvs::srv::SetMap::Request> const request,
+        std::shared_ptr<example_srvs::srv::SetMap::Response> response)>;
+
+    // Define path generation service callback type
+    using GeneratePathServiceCallback = std::function<void(
+        std::shared_ptr<example_srvs::srv::GetPath::Request> const request,
+        std::shared_ptr<example_srvs::srv::GetPath::Response> response)>;
+    /**
+     * @brief Ensure public dtor is virtual
+     * @note
+     * http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-virtual
+     */
+    virtual ~MiddlewareHandle() = default;
+
+    /**
+     * @brief Register a callback for the set map service
+     * @param callback The callback to register
+     */
+    virtual void register_set_map_service(SetMapServiceCallback callback) = 0;
+
+    /**
+     * @brief Register a callback for the generate path service
+     * @param callback The callback to register
+     */
+    virtual void register_generate_path_service(
+        GeneratePathServiceCallback callback) = 0;
+
+    /**
+     * @brief Log error message
+     * @param msg The message to log
+     */
+    virtual void log_error(std::string const& msg) = 0;
+
+    /**
+     * @brief Log info message
+     * @param msg The message to log
+     */
+    virtual void log_info(std::string const& msg) = 0;
+  };
+
+  /**
+   * @brief Builds paths given a map
+   * @param mw The middleware handle for interacting with services
+   */
+  PathGenerator(std::unique_ptr<MiddlewareHandle> mw) : mw_{std::move(mw)} {
```

Comment:
If the middleware object in the original scope is accessed it will be UB... or it will throw, not sure.

The `std::move` is required because it's a `unique_ptr`. I don't think it will compile otherwise.